<a href="https://colab.research.google.com/github/alexDx12/gb_statistics/blob/lesson_2/lesson_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Задачи 1 и 2

__Постановка задачи__

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Случайная величина X задаёт число правильных ответов при простом угадывании. Найдите математическое ожидание данной случайной величины. Найдите дисперсию случайной величины X.  
Подсказка: постройте закон распределения случайной величины X.

__Решение__

Исходя из формулировки задачи, речь идёт о распределении случайной величины по биномиальному закону - вероятность получения определённого количества правильных ответов (k) на определённое количество вопросов (n), при определённой вероятности угадать правильный ответ (p).  
В результате, вероятность Px случайной величины X получения k правильных ответов на n вопросов определяется следующей формулой:  
 
$$Px(X = k) = C_n^k p^k (1 - p)^{n - k}$$

Импорт библиотек и скриптов:

In [1]:
import numpy as np
from math import factorial 

Список вариантов числа правильных ответов контрольной работы (k_values):

In [2]:
k_values = [i for i in range(6)]
k_values

[0, 1, 2, 3, 4, 5]

Общее количество вопросов (n):

In [3]:
n = 5

Вероятность получить правильный ответ на каждый из вопросов (p):

In [4]:
p = 1 / 4
p

0.25

Список вариантов вероятностей для каждого варианта числа правильных ответов контрольной работы (px_values):

In [5]:
Px_values = []
for i in range(6):
    Px = (factorial(n) / (factorial(k_values[i]) * factorial(n - k_values[i]))) * p**k_values[i] * (1 - p)**(n - k_values[i])
    Px_values.append(Px)
Px_values

[0.2373046875,
 0.3955078125,
 0.263671875,
 0.087890625,
 0.0146484375,
 0.0009765625]

Преобразование списков k_values и px_values в массивы numpy для удобства дальнейших вычислений:

In [6]:
k_values = np.array(k_values)
Px_values = np.array(Px_values)

Математическое ожидание числа правильных ответов кнтрольной работы (m):

In [7]:
m = (k_values * Px_values).sum()
m

1.25

Дисперсия числа правильных ответов контрольной работы (d):

In [8]:
d = (((k_values - m) ** 2) * Px_values).sum()
d

0.9375

Ответ: математическое ожидание случайной величины равно 1.25; дисперсия случайной величины равна 0.9375.

### Задача 3

__Постановка задачи__

Пользователь получает в среднем 10 писем со спамом на свой почтовый ящик за сутки. Найти число N, такое, что с вероятностью 0.95 пользователь получит не более N писем со спамом за текущий день.

__Решение__

Исходя из формулировки задачи, речь идёт о распределении случайной величины по формуле Пуассона - вероятность получения определённого количества писем со спамом (N) с определённой интенсивностью (lambda_).  
В результате, вероятность p случайной величины X получения N писем со спамом определяется следующей формулой:  
 
$$P(X = N) = \dfrac{\lambda^N e^{-\lambda}}{N!}$$
 
Чтобы определить количество писем со спамом не более N, необходимо сложить отдельные вероятности получения количества писем от 0 до N.

Импорт библиотек и скриптов:

In [9]:
import numpy as np

Вспомогательные классы и функции:

In [17]:
def poisson_proba(N: int, lambda_: float) -> float:
    """Формула Пуассона.
    """
    
    return (lambda_ ** N) * (np.exp(-lambda_)) / np.math.factorial(N)

Суммирование вероятностей получения писем со спамом до тех пор, пока суммарная вероятность не достигнет 0.95:

In [18]:
N = 0
lambda_ = 10
p = 0
while True:
    p += poisson_proba(N=N, lambda_=lambda_)
    if p >= 0.95:
      break
    N += 1
p

0.9512595966960214

In [19]:
N

15

Ответ: с вероятностью 0.95 пользователь получит не более 15 писем со спамом в день.

### Задача 4

__Постановка задачи__

Производятся выстрелы по мишени. Вероятность попадания в мишень при одном выстреле равна 0.01. Сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 6 раз будет совершено попадание?
Подсказка: 1) "Вероятность попасть k раз при n выстрелах" - на какое распределение это похоже? 2) А если нам нужна вероятность P(X >= k), а не P(X = k)? 3) Здесь предстоит немножко покодить.

__Решение__

Исходя из формулировки задачи, речь идёт о распределении случайной величины по биномиальному закону - вероятность попадания определённого количества раз в мишень (k) с помощью определённого количества выстрелов (n), при определённой вероятности попадания (p).  
В результате, вероятность Px случайной величины X попадания k раз в мишень с помощью n выстрелов определяется следующей формулой:  
 
$$Px(X = k) = C_n^k p^k (1 - p)^{n - k}$$
 
Чтобы определить количество выстрелов для попадания как минимум k раз в мишень, необходимо складывать отдельные вероятности попадания от 0 до n до тех пор, пока суммарная вероятность не достигнет 0.9.

Импорт библиотек и скриптов:

In [20]:
import numpy as np

Вспомогательные классы и функции:

In [21]:
def binominal_proba(k: int, n: int, p: float) -> float:
    """Формула биномиального распределения.
    """
    
    return (np.math.factorial(n) / (np.math.factorial(k) * np.math.factorial(n - k))) * (p ** k) * (1 - p) ** (n - k)  

Суммирование вероятностей попадания по мишени до тех пор, пока суммарная вероятность не достигнет 0.9:

In [22]:
k = 6
n = 6
p = 0.01
Px = 0
while True:
    Px += binominal_proba(k, n, p)
    if Px >= 0.9:
      break
    n += 1
Px

0.9185129699543314

In [23]:
n

230

Ответ: с вероятностьяю 0.9 необходимо сделать 230 выстрелов, чтобы быть уверенным, что хотя бы 6 раз будет совершено попадание.